### PR TITLE
#2219 - Review dispatch on numeric type in concrete intersection

### PR DESCRIPTION
**Pull request by mforets**
See https://github.com/JuliaReach/LazySets.jl/issues/2219.

### CODE REVIEW

File: docs/src/lib/binary_functions.md [hunk 1]
```diff
@@ -64,26 +64,26 @@ monotone_chain!
...
-intersection(::AbstractPolyhedron{N}, ::AbstractPolyhedron{N}) where {N<:Real}
+intersection(::AbstractSingleton, ::LazySet)
+intersection(::Line2D, ::Line2D)
+intersection(::AbstractHyperrectangle, ::AbstractHyperrectangle)
+intersection(::Interval, ::Interval)
+intersection(::Interval, ::HalfSpace)
+intersection(::Interval, ::Hyperplane)
+intersection(::Interval, ::LazySet)
+intersection(::AbstractHPolygon, ::AbstractHPolygon, ::Bool=true)
+intersection(::AbstractPolyhedron{N}, ::AbstractPolyhedron{N}) where {N}
```
Comment (mforets) [inline docs/src/lib/binary_functions.md:75]: the reason i kept `N` in this method is that it requires a  `_default_...(N)`. we can probably use the promotion there too.

File: src/ConcreteOperations/intersection.jl [hunk 2]
```diff
@@ -448,30 +417,14 @@ function intersection(X::Interval{N}, Y::LazySet{N}
         return EmptySet{N}(1)
     end
 end
-
 # symmetric method
-function intersection(Y::LazySet{N}, X::Interval{N}
-                     ) where {N<:Real}
-    return intersection(X, Y)
-end
+intersection(Y::LazySet, X::Interval) = intersection(X, Y)
```
Comment (schillic) [inline src/ConcreteOperations/intersection.jl:421]:
```suggestion
intersection(Y::LazySet, X::Interval) = _intersection_interval(X, Y)
```

File: src/ConcreteOperations/intersection.jl [hunk 3]
```diff
@@ -448,30 +417,14 @@ function intersection(X::Interval{N}, Y::LazySet{N}
...
-end
-function intersection(S::AbstractSingleton{N}, X::Interval{N}
-                     ) where {N<:Real}
-    return invoke(intersection, Tuple{typeof(S), LazySet{N}}, S, X)
-end
+# disambiguations
+intersection(X::Interval, H::AbstractHyperrectangle) = _intersection_interval(X, H)
+intersection(H::AbstractHyperrectangle, X::Interval) = _intersection_interval(X, H)
+intersection(X::Interval, S::AbstractSingleton) = _intersection_interval(X, S)
+intersection(S::AbstractSingleton, X::Interval) = _intersection_interval(X, S)
```
Comment (schillic) [inline src/ConcreteOperations/intersection.jl:427]: The old version used the `Singleton` method and I think this is is the more specific type.

File: src/ConcreteOperations/intersection.jl [hunk 4]
```diff
@@ -950,38 +859,32 @@ A polyhedron.
...
-function intersection(rm::ResetMap{N}, P::AbstractPolyhedron{N}) where {N<:Real}
-    return intersection(P, rm)
-end
+intersection(rm::ResetMap, P::AbstractPolyhedron) = intersection(P, rm)
 
 # more efficient version for polytopic
-function intersection(P::AbstractPolyhedron{N},
-                      rm::ResetMap{N, <:AbstractPolytope}) where {N<:Real}
+function intersection(P::AbstractPolyhedron{M},
+                      rm::ResetMap{N, <:AbstractPolytope}) where {M, N}
```
Comment (schillic) [inline src/ConcreteOperations/intersection.jl:871]: The `M` is not used (same in the symmetric method below).

File: src/ConcreteOperations/intersection.jl (hunk 3 above)
Comment (mforets) [reply src/ConcreteOperations/intersection.jl:427]: :+1:  i change this to `_intersection_singleton`

File: src/ConcreteOperations/intersection.jl (hunk 4 above)
Comment (mforets) [reply src/ConcreteOperations/intersection.jl:871]: :+1:

we'll reduce a bunch of LOCs with the `@commutative` macro from https://github.com/JuliaReach/LazySets.jl/pull/2286  (it is held back by incompability with julia v1.0 )

**Review by schillic — APPROVED** [review]

**Pull request merged**